src/main/java/org/alfresco/extensions/bulkexport/controler/Engine.java

```diff
@@ -1,230 +1,229 @@
+
+    /** How many Nodes are exported per process*/
+    private int exportChunkSize;
+
+    /**
+     * Engine Default Builder
+     *
+     * @param dao Data Access Object
+     * @param fileFolder File and Folder magager
+     */
+    public Engine(AlfrescoExportDao dao, FileFolder fileFolder, boolean exportVersions, boolean revisionHead, boolean useNodeCache, int nbOfThreads, int exportChunkSize) {
+        this.dao = dao;
+        this.fileFolder = fileFolder;
+        this.exportVersions = exportVersions;
+        this.revisionHead = revisionHead;
+        this.useNodeCache = useNodeCache;
+        this.nbOfThreads = nbOfThreads;
+        this.exportChunkSize = exportChunkSize;
+    }
+
+    /**
+     * Recursive method to export alfresco nodes to file system 
+     *
+     * @param nodeRef
+     */
+    public void execute(NodeRef nodeRef) throws Exception {
+        // case node is folder create a folder and execute recursively 
+        // other else create file 
+        log.debug("execute (noderef)");
+
+        if (!this.dao.isNodeIgnored(nodeRef.toString())) {
+            log.info("Find all nodes to export (no history)");
+            List<NodeRef> allNodes = getNodesToExport(nodeRef);
+            log.info("Nodes to export = " + allNodes.size());
+            exportNodes(allNodes);
+        }
+        log.debug("execute (noderef) finished");
+    }
+
+    private List<NodeRef> getNodesToExport(NodeRef rootNode) throws Exception {
+        List<NodeRef> nodes = null;
+        if (useNodeCache) {
+            nodes = retrieveNodeListFromCache(rootNode);
+        }
+
+        if (nodes == null) {
+            nodes = findAllNodes(rootNode);
+            storeNodeListToCache(rootNode, nodes);
+            if (useNodeCache) {
+                log.info("Generated Cached Node list");
+                throw new CacheGeneratedException("Generated Cached Node List Only");
+            }
+        } else {
+            log.info("Using Cached Node list");
+        }
+
+        return nodes;
+    }
+
+    private String nodeFileName(NodeRef rootNode) {
+        File fname = new File(fileFolder.basePath(), rootNode.getId() + ".cache");
+        return fname.getPath();
+    }
+
+    private void storeNodeListToCache(NodeRef rootNode, List<NodeRef> list) throws Exception {
+        // get a better name
+        FileOutputStream fos = new FileOutputStream(nodeFileName(rootNode));
+        ObjectOutputStream oos = new ObjectOutputStream(fos);
+        oos.writeObject(list);
+        oos.close();
+        fos.close();
+    }
+
+    private List<NodeRef> retrieveNodeListFromCache(NodeRef rootNode) throws Exception {
+        List<NodeRef> list = null;
+
+        try {
+            FileInputStream fis = new FileInputStream(nodeFileName(rootNode));
+            ObjectInputStream ois = new ObjectInputStream(fis);
+            list = (List<NodeRef>) ois.readObject();
+            ois.close();
+        } catch (FileNotFoundException e) {
```

Comment:
Maybe create a method to check if the file exist instead of just throwing an exception